### PR TITLE
add foreign key support for users id and posts id 

### PR DESCRIPTION
**Pull request by madtibo**
This option is using the "--foreign-keys" switch

WARNING: when using the foreign keys option, some entries in votes and postlinks might be updated to enforce data integrity

### CODE REVIEW

File: sql/PostLinks_fk.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+-- impossible to enforce so set NULL
```
Comment (musically-ut) [inline sql/PostLinks_fk.sql]: This seems a bit too harsh. 

I think we should err on the side of caution and let data remain in the DB and not add the foreign key constraints here. Another option is to add explicit option for `--hard-foreign-keys` which will add these foreign key constraints while dropping data.

What do you think?

File: sql/Tags_fk.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+-- dummy query
+SELECT 1;
```
Comment (musically-ut) [inline sql/Tags_fk.sql:2]: Is this here because Postgres complains if we try to run just a file with a comment?

File: sql/Votes_fk.sql [hunk 3]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE Votes ADD CONSTRAINT fk_votes_userid FOREIGN KEY (userid) REFERENCES users (id);
+-- impossible to enforce so set NULL
+UPDATE Votes SET postid=NULL WHERE postid NOT IN (SELECT DISTINCT id FROM Posts);
```
Comment (musically-ut) [inline sql/Votes_fk.sql]: Same comment as before about erring in the side of caution and keeping extra data in the DB.

File: sql/Votes_pre.sql [hunk 4]
```diff
@@ -1,7 +1,7 @@
 DROP TABLE IF EXISTS Votes CASCADE;
 CREATE TABLE Votes (
    Id                int         PRIMARY KEY ,
-   PostId            int         not NULL    ,
+   PostId            int         , -- not NULL    ,
```
Comment (musically-ut) [inline sql/Votes_pre.sql]: I'd rather not add the foreign key constraints if they require dropping the `not NULL` on these fields.

One will not be updating the tables anyway after they have been created once. So the primary reason to have foreign key constraints is for performance (to make `EXISTS` queries faster, at least), correct?

In that case, I'd rather leave them up to the discretion of the user while preserving as many 'sanity' checks as we can (e.g. not NULL constraints) which are provided by the underlying real-data.

**Review by musically-ut — COMMENTED** [review]
Thanks for the feature add!

I am completely behind adding foreign key constraints for performance reason. However, I have some reservations about changing the underlying data, esp. about dropping data which _may_ have special meaning for people. 

I would be happy to accept the PR in two parts: adding the Foreign Keys which are _safe_, and then we can discuss the data-altering ones separately.

What do you think?

File: sql/Tags_fk.sql (hunk 2 above)
Comment (madtibo) [reply sql/Tags_fk.sql:2]: yes, that is the reason.

**Comment by madtibo** [comment]
The foreign key are useful for the planer to optimise queries. But constraints are mainly important to enforce the data business integrity. A NULL value for a foreign key means the linked entry has been deleted. I would think that data are more 'sane' when there is no link to non-existent data. But it is just a point of view :-)

I chose to nullify some values to keep the maximum of the information. But the drawback is that I had to drop the NOT NULL constraint.  

An other option to keep the not null constraint is to delete the offending rows before creating the foreign keys. That would be more damaging on the data so definitely, a change to '--hard-foreign-keys' would be a good idea.

@musically-ut Tell me what you prefer, I can make the changes.

**Comment by musically-ut** [comment]
Okay, in this case, let's split it into two separate PRs: one with `--foreign-keys` (safe) and another with `--hard-foreign-keys`. Then in this version, we'll only add foreign-key constraints which are safe to add.

After this, I think #9 would be a great feature to add!

Thanks again!

**Comment by madtibo** [comment]
OK, I'll look into it to create and use these options.

**Comment by madtibo** [comment]
@musically-ut I removed the data update. The constraint are set as `not valid`, which means the constraints will not be verified for current data. They will be enforced on new inserts and updates.
I put some comments on how to enforce the constraints validation in the _Votes_fk.sql_ and _PostLinks_fk.sql_ files.

What do you think of this version?

File: load_into_pg.py [hunk 5]
```diff
@@ -95,13 +203,12 @@ def handleTable(table, keys, insertJson, dbname, mbDbFile, mbHost, mbPort, mbUse
...
                                             ]
                                         )
-
                             if len(valuesStr) > 0:
                                 cmd = 'INSERT INTO ' + table + \
                                       ' VALUES\n' + valuesStr + ';'
                                 cur.execute(cmd)
                                 conn.commit()
-                        six.print_('Table processing took {:.1f} seconds'.format(time.time() - start_time))
+                        six.print_('Table processing took {1:.1f} seconds'.format(table, time.time() - start_time))
```
Comment (musically-ut) [inline load_into_pg.py]: Why not 'Processing table {} took ...' ?

**Review by musically-ut — APPROVED** [review]

File: load_into_pg.py (hunk 5 above)
Comment (madtibo) [reply load_into_pg.py]: I might have had the idea to log it. Well spotted!
I added it.

**Comment by musically-ut** [comment]
LGTM! Thanks!

**Comment by madtibo** [comment]
Great! Let's go for the third PR...

**Pull request merged**